lib/formats/csv.flow

```diff
@@ -3,23 +3,121 @@ import ds/array;
+
+	ParseCsvOption ::= ParseCsvSeparator, ParseCsvDetectSeparator, ParseCsvRectangle;
+		ParseCsvSeparator(sep : string); // any character except double quotes or empty string
+		ParseCsvDetectSeparator(); // try to detect if separator is either of , ; \t 
+		ParseCsvRectangle(); // treat inconsistent number of columns as an error
+
+	// By default uses ',' as a separator.
+	// If an error occurs the second element of the pair will contain an error message
+	// and first element of the pair may contain a partial result.
+	parseCsvData(data : string, styles : [ParseCsvOption]) -> Pair<[[string]], string>;
+}
+
+compiledGrammarsR : ref Tree<string, [PegOp]> = ref makeTree();
+
+parseCsvData(data : string, styles : [ParseCsvOption]) -> Pair<[[string]], string> {
+	styleSepM = tryExtractStruct(styles, ParseCsvSeparator(""));
+	sepStr = eitherFn(styleSepM, \styleSep -> styleSep.sep, \ -> {
+		if (containsStruct(styles, ParseCsvDetectSeparator())) {
+			sep = detectCsvSeparator(data).second;
+			devtrace("parseCsvData: detected separator is " + sep);
+			sep;
+		} else {
+			","; // default separator
+		}
+	});
+
+	// convert string to unicode escapes
+	sep = concatStrings(map(s2a(sepStr), \code -> {
+		hex = formatHex(code);
+		substring("\\u0000", 0, 6 - strlen(hex)) + hex;
+	}));
+
+	// cannot have double quotes or empty string as a separator
+	if (strContains(sep, "\\u0022") || sepStr == "") {
+		Pair([[]], "invalid separator");
+	} else {
+		grammarM = lookupTree(^compiledGrammarsR, sep);
+		grammar = eitherFn(grammarM, idfn, \ -> {
+			newOps = compilePegGrammar(formatString("#include formats/csv/csv.lingo-template", [sep]));
+			compiledGrammarsR := setTree(^compiledGrammarsR, sep, newOps);
+			newOps;
+		});
+
+		actions = SemanticActions(setTree(defaultPegActions.t, "unescapeQ", \args : [flow] -> {
+			str = flow2s(args[0]);
+			trimmed = substring(str, 1, strlen(str) - 2);
+			strReplace(trimmed, "\"\"", "\"");
+		}));
+		triple = parsic3(grammar, data, actions, [[]]);
+
+		if (triple.third != "") {
+			devtrace("parseCsvData: " + triple.third);
+			Pair([[]], "parsing error");
+		} else {
+			// trim empty rows at the end
+			result = triple.first;
+			idxM = lastfindi(result, \row -> row != [""]);
+			trimmedResult = eitherMap(idxM, \idx -> take(result, idx + 1), result);
+
+			if (containsStruct(styles, ParseCsvRectangle())) {
+				numCols = uniq(map(trimmedResult, length));
+				if (length(numCols) > 1) {
+					Pair(trimmedResult, "inconsistent number of columns");
+				} else {
+					Pair(trimmedResult, "");
+				}
+			} else {
+				Pair(trimmedResult, "");
+			}
+		}
+	}
+}
+
+detectCsvSeparator(s : string) -> Pair<int, string> {
+	s_part = strLeft(s, 2000);
+	count_cm = if (strCountOf(s_part, ",") > 0) getBestRowCount(s_part, ",") else 0;
+	count_sc = if (strCountOf(s_part, ";") > 0) getBestRowCount(s_part, ";") else 0;
+	count_tb = if (strCountOf(s_part, "\t") > 0) getBestRowCount(s_part, "\t") else 0;
```

Comment:
Avoid copy-paste here, you could define delimeters as an array of chars or a string and iterate over.